### ISSUE TITLE
Unfriendly Error Messages When Creating An Index

### ISSUE THREAD

**Issue by chamook** [elastic/elasticsearch-net#2985]
When creating an index through Nest there are a bunch of things that can be configured incorrectly that prevent the index from being created. I think it would be useful to have the message in the exception thrown be more detailed and helpful about how to go about fixing the problem.

This is a small example of one such issue that caused me trouble recently:

```fsharp
let createClient (uri : Uri) username password =
    let connection = new ConnectionSettings(uri)
    connection
        .BasicAuthentication(username, password)
        .ThrowExceptions()
        .DisableDirectStreaming()
    |> Transport<IConnectionSettingsValues>
    |> ElasticClient
    :> IElasticClient

let createIndexBroken () =
    let client = createClient (Uri "http://localhost:9200") "" ""

    let c = CustomAnalyzer()
    c.Tokenizer <- "standard"
    c.Filter <- [| "ascii_folding" |] //this shouldn't have the _

    let analyzers = 
        c
        :> IAnalyzer
        |> fun x -> ("test-analyzer", x)
        |> Array.singleton
        |> dict
        |> Analyzers
        :> IAnalyzers

    let a = Analysis()
    a.Analyzers <- analyzers

    let createRequest =
        CreateIndexRequest("test" |> IndexName.op_Implicit, IndexState())

    let settings = IndexSettings()
    settings.Analysis <- a
    createRequest.Settings <- settings

    client.CreateIndexAsync(createRequest :> ICreateIndexRequest)
    |> Async.AwaitTask
    |> Async.RunSynchronously
```

The value being set for `filter` shouldn't contain an underscore. This causes the whole thing to fail with the following exception:

```
Unhandled Exception: System.AggregateException: One or more errors occurred. (Could not complete the request to Elasticsearch.) ---> Elasticsearch.Net.ElasticsearchClientException: Could not complete the request to Elasticsearch.
   at Elasticsearch.Net.RequestPipeline.BadResponse[TReturn](ElasticsearchResponse`1& response, RequestData data, List`1 pipelineExceptions) in C:\Users\russ\source\elasticsearch-net-5.x\src\Elasticsearch.Net\Transport\Pipeline\RequestPipeline.cs:line 467
   at Elasticsearch.Net.Transport`1.<RequestAsync>d__15`1.MoveNext() in C:\Users\russ\source\elasticsearch-net-5.x\src\Elasticsearch.Net\Transport\Transport.cs:line 171
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Nest.ElasticClient.<Nest-IHighLevelToLowLevelDispatcher-DispatchAsync>d__261`4.MoveNext()
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.FSharp.Control.AsyncBuilderImpl.commit[a](AsyncImplResult`1 res)
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronouslyInCurrentThread[a](CancellationToken token, FSharpAsync`1 computation)
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronously[a](CancellationToken token, FSharpAsync`1 computation, FSharpOption`1 timeout)
   at Microsoft.FSharp.Control.FSharpAsync.RunSynchronously[T](FSharpAsync`1 computation, FSharpOption`1 timeout, FSharpOption`1 cancellationToken)
   at Program.main(String[] argv) in /Users/adam/Development/search-indexer/src/Storytel.Indexer/Program.fs:line 123
Abort trap: 6
```

Which is not very illuminating. Would it be possible to improve the message returned here, such that it gave some information about the value for `Filter` being set incorrectly?

NEST and ElasticSearch.Net version 5.5.0
Using ElasticSearch in a docker container locally (`docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:6.1.1`)

**Comment by russcam** [comment]
There's quite a bit of information available on [`ElasticsearchClientException`](https://github.com/elastic/elasticsearch-net/blob/6961b8df998c53bb86a0da0613c8d167b9714327/src/Elasticsearch.Net/Exceptions/ElasticsearchClientException.cs) which can be inspected to ascertain the reason:

- the [Request](https://github.com/elastic/elasticsearch-net/blob/6961b8df998c53bb86a0da0613c8d167b9714327/src/Elasticsearch.Net/Exceptions/ElasticsearchClientException.cs#L12) (Request bytes will be available if `.DisableDirectStreaming()` is set)
- the [Response](https://github.com/elastic/elasticsearch-net/blob/6961b8df998c53bb86a0da0613c8d167b9714327/src/Elasticsearch.Net/Exceptions/ElasticsearchClientException.cs#L14) (Response bytes will be available if `.DisableDirectStreaming()` is set)
- [An audit trail](https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/audit-trail.html) for the request
- The failure reason
- [Debug Information](https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/debug-information.html) containing a friendly description of all of the above

With `.DisableDirectStreaming()` set on `ConnectionSettings`, the response contains the JSON response from Elasticsearch, which is

```json
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]"}],"type":"illegal_argument_exception","reason":"Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]"},"status":400}
``` 
Since `.ThrowExceptions()` is enabled in the example, and the response from Elasticsearch is a 400 Bad Request, the client will throw an `ElasticsearchClientException`.

With the client configured not to throw exceptions (the default), a 400 Bad Request will not cause the client to throw an exception. A _valid_ response can be ascertained by inspecting the `.IsValid` property; if the property is not valid, and Elasticsearch has returned an error JSON object, the error will be deserialized into the `.ServerError` property:

```fsharp
    let response = client.CreateIndexAsync(createRequest :> ICreateIndexRequest)
                    |> Async.AwaitTask
                    |> Async.RunSynchronously
    
    if response.IsValid = false then response.ServerError.Dump() // using LINQPad
```
<table id="t1"><tr><td class="typeheader" colspan="2">ServerError</td></tr><tr id="sum2"><td colspan="2" class="summary">ServerError: 400Type: illegal_argument_exception Reason: "Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]"</td></tr><tr><th class="member" title="Elasticsearch.Net.Error">Error</th><td><table id="t3"><tr><td class="typeheader" colspan="2">Error</td></tr><tr id="sum4"><td colspan="2" class="summary">Type: illegal_argument_exception Reason: "Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]"</td></tr><tr><th class="member" title="Elasticsearch.Net.ErrorCause">CausedBy</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyDictionary&lt;System.String,System.String&gt;">Headers</th><td><table id="t5"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="Elasticsearch.Net.ErrorCause+ErrorCauseMetadata">Metadata</th><td><table id="t6"><tr><td class="typeheader" colspan="2">ErrorCauseMetadata</td></tr><tr id="sum7"><td colspan="2" class="summary">Elasticsearch.Net.ErrorCause+ErrorCauseMetadata</td></tr><tr><th class="member" title="System.Int64?">BytesLimit</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int64?">BytesWanted</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int32?">Column</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;Elasticsearch.Net.ShardFailure&gt;">FailedShards</th><td><table id="t8"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.Boolean?">Grouped</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Index</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">IndexUUID</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Language</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">LicensedExpiredFeature</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int32?">Line</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Phase</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;System.String&gt;">ResourceId</th><td><table id="t9"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.String">ResourceType</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Script</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;System.String&gt;">ScriptStack</th><td><table id="t10"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.Int32?">Shard</th><td><i>null</i></td></tr></table></td></tr><tr><th class="member" title="System.String">Reason</th><td>Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]</td></tr><tr><th class="member" title="IReadOnlyCollection&lt;Elasticsearch.Net.ErrorCause&gt;">RootCause</th><td><table id="t11"><tr><td class="typeheader" colspan="5">List&lt;ErrorCause&gt; (1 item)</td></tr><tr><th title="Elasticsearch.Net.ErrorCause">CausedBy</th><th title="Elasticsearch.Net.ErrorCause+ErrorCauseMetadata">Metadata</th><th title="System.String">Reason</th><th title="System.String">StackTrace</th><th title="System.String">Type</th></tr><tr><td><i>null</i></td><td><table id="t12"><tr><td class="typeheader" colspan="2">ErrorCauseMetadata</td></tr><tr id="sum13"><td colspan="2" class="summary">Elasticsearch.Net.ErrorCause+ErrorCauseMetadata</td></tr><tr><th class="member" title="System.Int64?">BytesLimit</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int64?">BytesWanted</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int32?">Column</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;Elasticsearch.Net.ShardFailure&gt;">FailedShards</th><td><table id="t14"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.Boolean?">Grouped</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Index</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">IndexUUID</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Language</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">LicensedExpiredFeature</th><td><i>null</i></td></tr><tr><th class="member" title="System.Int32?">Line</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Phase</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;System.String&gt;">ResourceId</th><td><table id="t15"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.String">ResourceType</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Script</th><td><i>null</i></td></tr><tr><th class="member" title="IReadOnlyCollection&lt;System.String&gt;">ScriptStack</th><td><table id="t16"><tr><td class="typeheader" colspan="1">(0 items)</td></tr></table></td></tr><tr><th class="member" title="System.Int32?">Shard</th><td><i>null</i></td></tr></table></td><td>Custom Analyzer [test-analyzer] failed to find filter under name [ascii_folding]</td><td><i>null</i></td><td>illegal_argument_exception</td></tr></table></td></tr><tr><th class="member" title="System.String">StackTrace</th><td><i>null</i></td></tr><tr><th class="member" title="System.String">Type</th><td>illegal_argument_exception</td></tr></table></td></tr><tr><th class="member" title="System.Int32">Status</th><td>400</td></tr></table>

The client does not contain logic to determine if `ascii_folding` is a known token filter because it could be the name of a custom token filter defined within the the index. The general convention where strings are used for a setting is for the client to defer to Elasticsearch to determine if it's valid.

One more thing, since you're running against Elasticsearch 6.1.1, you should use a NEST 6.x client; [Major versions of the client are compatible with major versions of Elasticsearch](https://github.com/elastic/elasticsearch-net#compatibility-matrix), so NEST 5.x will work with any Elasticsearch 5.x. NEST 6.x has not GA'ed yet, so the [latest available version on nuget is 6.0.0-beta1](https://www.nuget.org/packages/NEST/6.0.0-beta1). We're working on the GA release and are super close. If you'd like to take 6.0.0-beta1 for a spin, awesome! Please let us know if you encounter any issues.

**Comment by Mpdreamz** [comment]
Reopened this with #2986 

I agree with @chamook that we could do better here. As @russcam points out plenty of information is available but having to rely on `DisableDirectStreaming()` is not a viable solution in production. Having better exception messages here also help with logging where typically only `Message` and perhaps `StackTrace` make it out.

Have a look here: https://github.com/elastic/elasticsearch-net/blob/0d77f2def06f34b54daee55d1942b644812c5cc5/src/Tests/Reproduce/GithubIssue2985.cs

To see what the `exception.Message` would be now.
